### ISSUE TITLE
Xenial blocker: browser versions on devstack

### ISSUE THREAD

**Issue by nedbat** [openedx-unsupported/configuration#3487]
Installing a base devstack Vagrant box from jibsheet/xenial-devstack (commit ab86af2). It eventually fails with:
```
TASK: [browsers | install browser debian packages] ****************************
<127.0.0.1> ESTABLISH CONNECTION FOR USER: ubuntu
<127.0.0.1> REMOTE_MODULE command gdebi -nq /tmp/google-chrome-stable_30.0.1599.114-1_amd64.deb #USE_SHELL
<127.0.0.1> EXEC ssh -C -tt -vvv -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/Users/ned/.ansible/cp/ansible-ssh-%h-%p-%r" -o StrictHostKeyChecking=no -o Port=2222 -o IdentityFile="/src/vm/xenial_devstack/configuration/vagrant/base/devstack/.vagrant/machines/default/virtualbox/private_key" -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=30 127.0.0.1 /bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1479533800.07-236951796018319 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1479533800.07-236951796018319 && echo $HOME/.ansible/tmp/ansible-tmp-1479533800.07-236951796018319'
<127.0.0.1> PUT /var/folders/j2/gr3cj3jn63s5q8g3bjvw57hm0000gp/T/tmpxKIia7 TO /home/ubuntu/.ansible/tmp/ansible-tmp-1479533800.07-236951796018319/command
<127.0.0.1> EXEC ssh -C -tt -vvv -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/Users/ned/.ansible/cp/ansible-ssh-%h-%p-%r" -o StrictHostKeyChecking=no -o Port=2222 -o IdentityFile="/src/vm/xenial_devstack/configuration/vagrant/base/devstack/.vagrant/machines/default/virtualbox/private_key" -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=30 127.0.0.1 /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=aivcbkmlbglufcyjajzjfxlvfebwpkvl] password: " -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-aivcbkmlbglufcyjajzjfxlvfebwpkvl; LANG=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/bin/python /home/ubuntu/.ansible/tmp/ansible-tmp-1479533800.07-236951796018319/command; rm -rf /home/ubuntu/.ansible/tmp/ansible-tmp-1479533800.07-236951796018319/ >/dev/null 2>&1'"'"''
failed: [default] => (item={'url': 'https://s3.amazonaws.com/vagrant.testeng.edx.org/google-chrome-stable_30.0.1599.114-1_amd64.deb', 'name': 'google-chrome-stable_30.0.1599.114-1_amd64.deb'}) => {"changed": true, "cmd": "gdebi -nq /tmp/google-chrome-stable_30.0.1599.114-1_amd64.deb", "delta": "0:00:01.213163", "end": "2016-11-19 05:36:41.401256", "item": {"name": "google-chrome-stable_30.0.1599.114-1_amd64.deb", "url": "https://s3.amazonaws.com/vagrant.testeng.edx.org/google-chrome-stable_30.0.1599.114-1_amd64.deb"}, "rc": 1, "start": "2016-11-19 05:36:40.188093", "warnings": []}
stderr: This package is uninstallable
Dependency is not satisfiable: libgcrypt11 (>= 1.4.5)
<127.0.0.1> ESTABLISH CONNECTION FOR USER: ubuntu
<127.0.0.1> REMOTE_MODULE command gdebi -nq /tmp/firefox-mozilla-build_42.0-0ubuntu1_amd64.deb #USE_SHELL
<127.0.0.1> EXEC ssh -C -tt -vvv -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/Users/ned/.ansible/cp/ansible-ssh-%h-%p-%r" -o StrictHostKeyChecking=no -o Port=2222 -o IdentityFile="/src/vm/xenial_devstack/configuration/vagrant/base/devstack/.vagrant/machines/default/virtualbox/private_key" -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=30 127.0.0.1 /bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1479533801.41-105698092597378 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1479533801.41-105698092597378 && echo $HOME/.ansible/tmp/ansible-tmp-1479533801.41-105698092597378'
<127.0.0.1> PUT /var/folders/j2/gr3cj3jn63s5q8g3bjvw57hm0000gp/T/tmpbzOYGm TO /home/ubuntu/.ansible/tmp/ansible-tmp-1479533801.41-105698092597378/command
<127.0.0.1> EXEC ssh -C -tt -vvv -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/Users/ned/.ansible/cp/ansible-ssh-%h-%p-%r" -o StrictHostKeyChecking=no -o Port=2222 -o IdentityFile="/src/vm/xenial_devstack/configuration/vagrant/base/devstack/.vagrant/machines/default/virtualbox/private_key" -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=30 127.0.0.1 /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=pnakyzrvhbcuspykjiipqookteterlnw] password: " -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-pnakyzrvhbcuspykjiipqookteterlnw; LANG=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/bin/python /home/ubuntu/.ansible/tmp/ansible-tmp-1479533801.41-105698092597378/command; rm -rf /home/ubuntu/.ansible/tmp/ansible-tmp-1479533801.41-105698092597378/ >/dev/null 2>&1'"'"''
changed: [default] => (item={'url': 'https://s3.amazonaws.com/vagrant.testeng.edx.org/firefox-mozilla-build_42.0-0ubuntu1_amd64.deb', 'name': 'firefox-mozilla-build_42.0-0ubuntu1_amd64.deb'}) => {"changed": true, "cmd": "gdebi -nq /tmp/firefox-mozilla-build_42.0-0ubuntu1_amd64.deb", "delta": "0:00:03.226221", "end": "2016-11-19 05:36:44.751925", "item": {"name": "firefox-mozilla-build_42.0-0ubuntu1_amd64.deb", "url": "https://s3.amazonaws.com/vagrant.testeng.edx.org/firefox-mozilla-build_42.0-0ubuntu1_amd64.deb"}, "rc": 0, "start": "2016-11-19 05:36:41.525704", "stderr": "", "stdout": "Selecting previously unselected package firefox-mozilla-build.\n(Reading database ... 93684 files and directories currently installed.)\nPreparing to unpack .../firefox-mozilla-build_42.0-0ubuntu1_amd64.deb ...\nAdding 'diversion of /usr/bin/firefox to /usr/bin/firefox.ubuntu by firefox-mozilla-build'\nUnpacking firefox-mozilla-build (42.0-0ubuntu1) ...\nSetting up firefox-mozilla-build (42.0-0ubuntu1) ...\nProcessing triggers for mime-support (3.59ubuntu1) ...", "warnings": []}

FATAL: all hosts have already failed -- aborting
```

@benpatterson said:
> TL;DR is: Firefox: stick with what we are using now. Chrome: Depends on what Xenial has available.
> 
> Firefox
> We use a tar for our firefox version, and Selenium has also introduced the geckodriver which will be needed for testing against firefox. However, that's also controlled by versions. (I know the latest selenium requires geckodriver for the latest firefox. But I don't know if the latest firefox requires geckodriver, if that makes sense.)
> The easiest path here is to not-change the firefox version and just continue unwrapping a tar. If we can confirm that Xenial handles this and tests can run on it, then I think we'll be in a good position.
> 
> Chrome
> You might remember the debian repository for Chrome on Precise was causing problems with installs. We have a weird workaround (IIRC) where we pull down Chrome and then delete the repository reference in lists.d. I don't know what's available in Xenial, including whether or not the older version of Chrome we now use is on there. But, as far as Chrome is concerned, I think the best thing to do here is to use the proper repository to install the older version of Chrome we use today. That way we don't have to worry about updating the ChromeDriver.
> 
> However, if that version of Chrome is not available in Xenial, then I'd suggest taking the latest of both the browser and the driver. We'd need some intelligence on the browsers role to understand which operating system version we are on, along the lines of "if precise: use older stuff; else: use newer stuff".

We need to find new browser PPAs to install.

**Comment by benpatterson** [comment]
@nedbat let's start from a more fundamental point (that I am a bit ignorant of...)

Are Chrome and Firefox available for Xenial to begin with?

**Comment by nedbat** [comment]
@benpatterson I've never had to go find browser versions for linux before. There certainly seems to be a download for Chrome for Linux: https://www.google.com/chrome/browser/desktop/index.html
Firefox has them too: https://launchpad.net/ubuntu/xenial/+source/firefox

Again, I would love the assistance of people who have been down these paths.... :)

**Comment by benpatterson** [comment]
I think we may be missing each other; I'll send a note.

**Comment by gsong** [comment]
@benpatterson What’s the smallest test I can run to verify that the browsers are installed properly?

`paver test_bokchoy` takes forever!

/cc @feanil @nedbat

**Comment by benpatterson** [comment]
A good thumb-in-the-air is to pull down the bok-choy repo itself and run
tox. I believe you'll want to be sure that xvfb is running with a certain
screen setting, but that's defined in the browsers role. You'll also want
to set DISPLAY to ":1"...

**Comment by jibsheet** [comment]
In the past, JZ has tipped me to running the first accessibility test, since they're bokchoy underneath but tend to be faster.

**Comment by gsong** [comment]
@jibsheet I’m not that familiar with running the tests, can you tell me the exact command? `paver test_bokchoy …`?

**Comment by gsong** [comment]
I’ve been able to install the browsers successfully:

```
$ firefox --version
Mozilla Firefox 42.0
$ chromedriver --version
ChromeDriver 2.25.426924 (649f9b868f6783ec9de71c123212b908bf3b232e)
$ phantomjs --version
1.9.8
```

(Side note: I tested successfully through Firefox 46, with FF 45 installable directly via `apt-get`)

I’ve manually tested them with Selenium 2.53.1.

<details>
<summary>Test code</summary>

```python
#!/usr/bin/env python
from selenium import webdriver

drivers = {
    'Chrome': webdriver.Chrome(),
    'Firefox': webdriver.Firefox(),
    'PhantomJS': webdriver.PhantomJS(),
}


def test_scenario(driver):
    """Test the codepad.org code execution service."""

    # Go to codepad.org
    print('Accessing URL')
    driver.get('http://codepad.org')

    # Select the Python language option
    print('Clicking on Python')
    python_link = driver.find_elements_by_xpath(
        "//input[@name='lang' and @value='Python']"
    )[0]
    python_link.click()

    # Enter some text!
    print('Entering text')
    text_area = driver.find_element_by_id('textarea')
    text_area.send_keys("print 'Hello,' + ' World!'")

    # Submit the form!
    print('Submit form')
    submit_button = driver.find_element_by_name('submit')
    submit_button.click()

    # Make this an actual test. Isn't Python beautiful?
    print('Successful test')
    assert "Hello, World!" in driver.page_source

    # Close the browser!
    driver.quit()


if __name__ == '__main__':
    for name, driver in drivers.items():
        print("Using {}".format(name))
        test_scenario(driver)
```
</details>

However, `paver test_bokchoy` and `edx/bok-choy` tox tests are still failing:


```
browser.py                 312 INFO     Using local browser: firefox [Default is firefox]
browser.py                 289 INFO     Using default firefox profile
browser.py                 244 DEBUG    Failed to instantiate browser: Message: The browser appears to have exited before we could connect. If you specified a log_file in the FirefoxBinary constructor, check it for details.
```

1. Any pointers on next steps?
1. How do we do a similar low-level test with bok-choy?
1. What else is sitting between bok-choy and selenium?

**Comment by benpatterson** [comment]
Hmm if you just run Firefox, will it successfully start the browser?

It could be your DISPLAY needs to be set....

**Comment by jibsheet** [comment]
@gsong The docs cover running tests https://github.com/edx/edx-platform/blob/6d2468d0dcf782a7561ecc61946146d6fae2762f/docs/en_us/internal/testing.rst#running-bok-choy-accessibility-tests

**Comment by nedbat** [comment]
This has been resolved.
